### ISSUE TITLE
Replace Shortlink custom algorithm with a standard implementation

### ISSUE THREAD

**Issue by MidnightLightning** [openstreetmap/openstreetmap-website#4205]
### Problem

OpenStreetMap has a custom [Shortlink](https://wiki.openstreetmap.org/wiki/Shortlink) algorithm that is used to generate a URL link to a specific point on the map in a terse format. That algorithm being custom means no other mapping tool (that I've seen documented) uses it as a shortening method. Being custom means the development team needs to keep updated the algorithm code, and documentation for it. If it were switched to a known standard algorithm, it would allow more interoperability with other tools, and would allow pruning out that code from the website (allowing the website code to focus more on OSM-specific business logic).

### Description

Replace the "Short Link" UI logic to use a known standard algorithm. The current Shortlink algorithm is similar to [Plus Codes](https://plus.codes), and the Plus Codes system has been adopted by many other mapping and addressing systems, so replacing it with that algorithm via a Plus Codes library would allow simplification of what the OSM site needs to do to calculate/parse "Short Links" and make generating OSM "Short Links" from other applications easier (as many will already have a Plus Codes library pulled in).  Though I note there has been some reluctance to add Plus Codes to other areas of the application (search bar: #1807); if that system is not preferred, the Geohash, and Geohash-36 systems are also similar to the current Shortlink algorithm and could be considered as more possible alternatives.

**Comment by tomhughes** [comment]
You say that this would save us having to keep our custom code updated but what exactly would we ever need to update? As far as I know our code is essentially untouched since it was created - it works and there's not reason it shouldn't continue to work or to believe it will need significant maintenance.

**Comment by MidnightLightning** [comment]
@tomhughes It's good that "it works" and has required minimal maintenance, however I presume like many projects, you as maintainer strive to make it "work the best it can" rather than "just working"? If it's been untouched for years, it could stand to be at least re-evaluated to see if it's still the best solution for the user workflow trying to be supported?

Do you have metrics on how frequently that feature is used? I would assume the most common use for that feature is to share temporary links with people (Alice finds something on the map, copies the shortlink, and sends it to Bob. Bob, at some point in the near future will read Alice's message, follow the link and know what she was talking about. After a few days, that link will likely never be used/followed again). Another possible use-case is from other sites that want to cross-link to information and might have shortlink-formatted links embedded in their website. Can you tell from your metrics if people use that tool the first way (no "referrer" page and each link only visited a few times in a short time range and then never again), or the second (visit has a referring page, and has visits spread out over a long time). If it's primarily used in the first way, then changing the algorithm would have minimal impact.

Or do your metrics show that feature gets a lot of use at all? The non-shortlink URL format isn't _that_ long/verbose a format; do people mostly just use that?

**Comment by Zaczero** [comment]
In my opinion, the advantages of transitioning to the plus codes standard offer minimal benefits for OSM at this time. Such a shift would necessitate additional coding and ensuring changes are backward compatible to avoid disrupting old links. I'm inclined to think that this effort might not be justified. The current system is quite robust.

**Comment by mmd-osm** [comment]
If you want something more interoperable, you should take a look at Geo URI instead, which is right next to the short link feature on osm.org. It includes an additional "z=?" parameter for the current zoom level. Check RFC 5870 for more details.

**Comment by MidnightLightning** [comment]
> If you want something more interoperable, you should take a look at Geo URI instead, which is right next to the short link feature on osm.org. It includes an additional "z=?" parameter for the current zoom level. Check RFC 5870 for more details.

That's a good point that there's a standards-based link style already there as an option. So what is the benefit to the site having multiple different types of "shareable" links to the same map view. I think it's great to encourage cross-linking between mapping services, and having multiple types of links could help encourage that, but if the link format is not a widely-accepted standard, is it actually being used and worth having as part of the codebase? Even if it's not a notable maintenance effort to keep it in, removing it could help the overall compiled size of the application (potential minor performance gain), and encouraging other sites to transition off of proprietary tools helps increase the network-effect of related sites, no?

I think this is not a severe issue/bug/error that needs to be corrected right away, but as a software developer myself, I do think that evaluating "where can/should the codebase be reduced/trimmed down" is a good practice alongside adding new features. What do the current usage metrics show for which types of share links are being used most frequently? And are they from long-term links from other sites, or links that people seem to be emailing/messaging each other?

**Comment by gravitystorm** [comment]
We can't remove the code for shortlinks, since that will break every shortlink that has ever been generated before. So we are required to maintain that code, even if we generated short links with a different algorithm in future.

I'm not sure what advantage there would be in generating shortlinks with any other algorithm, since they are just URLs to this site. And would therefore duplicate the shortlinks that we already offer. So I can't see there being any advantage to e.g. letting users choose the algorithm.

In short, I agree with the above comments from other contributors that this idea doesn't bring us any gains in terms of code maintenance or otherwise.

**Comment by MidnightLightning** [comment]
While it is up to the maintainers of the project to make the final call on this, I just wanted to note two bits of feedback on the given rationale why this was closed:

> We are required to maintain that code, even if we generated short links with a different algorithm in future

The concept that once created a bit of code needs to be maintained **_forever_** is not one that most software packages follow. Software has features that get deprecated and removed as needs and behaviors shift. If a specific thing is designed to be a standard, then it is good for it to be maintained for a long time, but even then, the concept of a new standard deprecating an older one is not an unheard-of thing.

The "Shortlink" algorithm is documented on the wiki, but not presented as "a standard" in a way that other applications could generate on their own. Hence, I find it odd why it's not open to even consider deprecating it.

Secondly, I've noted several times that it would be helpful to see usage metrics on that feature to tell if it is being used and in what way. There's been declarative statements made that the focus is on supporting "every shortlink that has ever been generated", but no evidence given that old shortlinks are even used (my thought being that a typical use-case would be sharing a link with a friend, and once that friend has visited, that shortlink would be not used again). Is the maintaining team gathering metrics on site feature usage and using that to evaluate development options?
